readme.md
# gBlog

client ID-
895962107065-0du9jvr9dnv6nrlv9s54usgls5do7572.apps.googleusercontent.com
client secrect-
oN4bHjHZHlOHFfwF0Dj4iNID

![gBlog](gBlog.png)

Build a Full Stack app that allows users to CRUD blog posts.

<hr>
# Wait!!!!

Before you rush in, scan this whole document. Soak it in. Take your time.

<hr>
# Think Things Through Thoroughly

These instructions provide a general outline and leave a lot of room for interpretation. They may not list exactly everything you need to do to build this CRUD app.

Be smart, be pragmatic, be excellent to each other.

<hr>
# gBlog Relationships

gBlog consists of the following relationships:

* A website has many blog posts.

* A blog post has a Title (Required), Body (Required), Author Name (Required), and Creation time (Required).

* A blog post can have many comments.

* A comment has a Body (Required), and Author Name (Required) and Creation time (Required).

> Do not worry about authentication and authorization right now. Auth is a stretch goal. Get your blog post CRUD working and UI created FIRST.

<hr>
# Steps for Success

## 1. Think and Diagram FIRST

Create an entity relationship diagram. Do not move on until you have.

## 2. Project Setup

* Get the project directory setup (using the express generator, or from scratch using npm init, npm install etc.).

* Create a new repo on github and prepare it for the initial check in.

* Check in your code

* Take pictures of the entity relationship diagrams and check them in.

## 3. Create the Database and Tables

* Work together to create the database and tables using the entity relationship diagram drawn as a reference.

* You may use your preferred method for creating the database and tables (psql, pg, knex migrations, etc.).

* Check in your code

## 4. CRUD Blog posts

* Create the CRUD routes for blog posts.

* Your routes should only return json.

* Make sure to test a route using postman or ajax requests before moving on to the next route.

* Check in your code

## 5. CRUD Blog comments

* Create the CRUD routes for blog comments.

* Your routes should only return json.

* Make sure to test a route using postman or ajax requests before moving on to the next route.

* Check in your code

## 6. Mock Up

* Draw/Mockup each of the following pages before moving on. Be sure to take pictures of your mockups and store them in your github repo.

* You should have an exact idea of the interaction and navigation that will happen in your app before you write a single line of code.

* The website will consist of the following pages:

+ this equals my inserted tasks!
+ Nav Bar
1- update nav bar across the website make
2- make links work

var excerptArray = blogPostBody.match(/[^?!\.]+[?!\.]+/g);

* Home page:

 * A list of all blog posts sorted by creation time and a 3 sentence excerpt from each.
 + 2 steps remaining!
 1- only showing a 3 sentence excerpt
 2- sorting blog posts by timestamp(come back to this after you can create a blog post)

DONE! * A link next to each blog post that will take you to the blog post's page.
! done!
DONE! * A link to create a new blog post

* Blog page:

DONE! * Display a single blog post. Should show Title, Author, Body and creation time.


DONE! * A link to edit the blog post.
! done!

 * A link to delete the blog post.
! done!

! done! * Below the blog post, display a form that will allow a user to add a comment.


!DONE! * Below the comment form, display all comments for the given blog post.



!DONE! * Next to each comment, display a link to delete the comment.

!DONE! * Next to each comment, display a link to edit the comment.

DONE! (except for wiping form after submission- look at galvanize personnel for this info!)

 ! done!* Create Blog Page:

 * Display a form with input boxes for Title, Author, Body
! done!
 * A button to submit the creation of the blog post
! done!

* Edit Blog Page:

! done!  * Display a form with input boxes for Title, Author, Body


DONE!  * A button to submit the edit of the blog post


* Edit Comment Page

  * Display a form with input boxes for body, author

  * A button to submit the edit of the comment.

Take pictures of your mockups and check them into your repo.

## 7. Front-End

* NOTE: Your api should only be returning JSON, so you will need to make ajax requests to get the blog contents.

* NOTE: For now, you can put your HTML/CSS/JS files in the `public` folder in your app directory. Use the [express static middleware](http://expressjs.com/en/api.html#express.static) to serve up the folder (This is enabled by default when using the express generator).

* Use bootstrap or another style framework of your choice. Plan ahead and use a grid system for your layout.

* Focus on creating one page at a time, focus on layout/styles and then focus on ajax requests/DOM manipulation.

* After you have finished a page, switch focus areas for the creation of the next page.

* Work efficiently.  

* Commit your code often.

## 8. Deploy to Heroku

* Create a Heroku App

* Provision a Postgres Database on your Heroku App

* Make sure you are using Environment Variables to connect to the database.

* Deploy early. Deploy often.

# Stretch Tasks

## 1. User Authentication  

> Remember that your api should only return json and should not perform any redirects. You can still create a session or set a cookie on the request and use the authentication concepts you have learned in class.

> Token-based authentication _can_ be used instead of cookies. Read up on it [here](http://code.tutsplus.com/tutorials/token-based-authentication-with-angularjs-nodejs--cms-22543) to see if you're up for the challenge. Also take a look at the passport [jwt strategy](https://github.com/themikenicholson/passport-jwt).

* Re-factor your application to use user authentication.

* Create a users table to store your users.

* Change the author_name columns on the blog posts table and blog comments table to author_id and add a relationship to the users table.

* Restrict blog update/deletion to the author that created the blog post.

* Restrict comment deletion to the author of the post and the comment author.

* You may use passport local strategy or role your own local strategy.

## 2. Deploy to firebase

* At this point your DB and express app are hosted together on heroku

* Express is capable of serving static files, but static deployment services like firebase specialize in static content. An opinion about that [here](https://divshot.com/blog/hosting/dont-host-static-sites-on-heroku/);

* Create a separate github repo for your front end project.

* Check in your front end code.

* Update your public javascript files to make API requests to the heroku server. You will need to enable CORS in express for your firebase domain. For security, only enable cors for https://yourappname.firebaseapp.com NOT for *

* Check in your code

* Deploy your static site to firebase using the firebase cli tools.

* Remove the public folder from your heroku instance so that only the API is accessible.

## 3. Drink a beer (or other appropriate beverage)

* YOU DID IT

* YOU'RE A FULL STACK DEVELOPER

* YOU HAVE CODE RUNNING ON 2 to 3 SEPARATE SERVERS ALL COMMUNICATING WITH EACH OTHER AND ACTUALLY WORKING

* YOU ROCK!!!!!

![GOOD JOB](https://media.giphy.com/media/xTiTnnLkYTDWSOWSHK/giphy.gif)


history.txt
    1  atom~/.zshrc
    2  atom -v
    3  atom ~/.zshrc
    4  echo $PATH
    5  brew install git
    6  git --version
    7  git config --global user.name 'Paul Torres'
    8  git config --global user.email 'torrepaul@gmail.com'
    9  brew install node
   10  node -v
   11  node
   12  atom ~/Desktop/test.js
   13  node ~/Desktop/test.js
   14  rm ~/Desktop/test.js
   15  echo $HOME
   16  pwd
   17  cd
   18  mkdir galvanize && \\ncd galvanize && \\nmkdir unit-1 && mkdir unit-2 && mkdir unit-3 && mkdir unit-4 && mkdir projects && \\nls -la
   19  ~cd projects
   20  ~ cd projects
   21  history
   22  atom ~/Desktop/test.js
   23  history | grep "galvanize"
   24  ~/.history
   25  ~vivim.zsh_history
   26  clear
   27  vim i
   28  create cli_notes.md
   29  cli_notes.md
   30  ls
   31  ~ cd downloads
   32  ls
   33  cd downloads
   34  cd movies
   35  ~/
   36  ls
   37  downloads
   38  cd desktop
   39  ls
   40  cd applications
   41  ls
   42  ls -l
   43  ls- hl
   44  ls -hl
   45  man ls
   46  touch test.txt
   47  makedir test
   48  mkdir test
   49  mv test.txt text/
   50  mv test.txt text/.
   51  mv test.txt test
   52  cd test
   53  ls
   54  open .
   55  echo ~
   56  brew install tree
   57  tree ~
   58  cd /
   59  cd ~
   60  cd ..
   61  ls -l ~/Documents/../
   62  cd /
   63  cd..
   64  cd ~/Library/Keyboard\ Layouts/        Printers/\nContainers/                 PubSub/\nCookies/                    Safari/\nCoreData/                   Saved\ Application\ State/\nDictionaries/               Screen\ Savers/\nFavorites/                  Services/\nFontCollections/            Sharing/\nFonts/                      Sounds/\nGoogle/                     Spelling/\nGroup\ Containers/          Suggestions/\nIdentityServices/           SyncedPreferences/\nInput\ Methods/             Voices/\nInternet\ Plug-Ins/         WebKit/\nKeyboard/                   com.apple.nsurlsessiond/\nKeyboard\ Layouts/          iMovie/\nKeyboardServices/           iTunes/\npaulanthonytorres:// $ cd ~/Library/Keyboard\ Layouts/               [11:52:36]\nAssistant/                  Logs/\nAssistants/                 Mail/\nAudio/                      Maps/\nCaches/                     Messages/\nCalendars/                  Metadata/\nCallServices/               Passes/\nColorPickers/               PreferencePanes/\nColors/                     Preferences/\nCompositions/               Printers/\nContainers/                 PubSub/\nCookies/                    Safari/\nCoreData/                   Saved\ Application\ State/\nDictionaries/               Screen\ Savers/\nFavorites/                  Services/\nFontCollections/            Sharing/\nFonts/                      Sounds/\nGoogle/                     Spelling/\nGroup\ Containers/          Suggestions/\nIdentityServices/           SyncedPreferences/\nInput\ Methods/             Voices/\nInternet\ Plug-Ins/         WebKit/\nKeyboard/                   com.apple.nsurlsessiond/\nKeyboard\ Layouts/          iMovie/\nKeyboardServices/
   65  cd ~/D
   66  dc ~/L
   67  cd ~/l
   68  cd ~/L
   69  cd ~/
   70  cd~/D
   71  cd~/l
   72  cd ~/Desktop/Galvanize\ Pre\ work
   73  cd ~
   74  cd Galvanize Pre work/
   75  cd downloads
   76  cd downloads/spectacle
   77  cd ~/Projects/week01
   78  cd Projects
   79  cd ~
   80  ls
   81  cd documents
   82  ls
   83  cd downloads
   84  cd Downloads
   85  cd ~
   86  cd Downloads
   87  cd ~/Documents
   88  ls
   89  cd ~/Downloads
   90  ls
   91  cd ~ Music
   92  cd ~/Music
   93  cd ~/Downloads/Atom.app
   94  ls
   95  cd Contents
   96  ls
   97  cd/Library
   98  cd ~/Library
   99  cd..
  100  ls -l ~/Documents/../
  101  cd.
  102  cd .
  103  cd /
  104  cd ..
  105  cd ~/Library
  106  cd ~/Projects/week01
  107  mkdir Projects
  108  cd Projects
  109  mkdir week01
  110  ls
  111  cd Projects
  112  ls
  113  cd week01
  114  mkdir notebook
  115  pwd
  116  ls
  117  touch notes.txt
  118  ls notes.txt
  119  cat notes.txt
  120  touch cli_notes.md
  121  cd cli_notes.md
  122  ls
  123  cd notes
  124  cd..
  125  ls
  126  cd week01
  127  cd week01/
  128  ls
  129  pwd
  130  cd notes.txt
  131  cat notes.txt
  132  ls -l notes.txt
  133  man notes.txt
  134  rm notes.txt
  135  history > notes.txt
  136  ls
  137  cat  notes.txt
  138  less notes.txt
  139  history > notes.txt 
  140  less notes.txt
  141  history>> notes.txt
  142  history >> notes.txt
  143  vim cli_notes.md
  144  cat cli_notes.md
  145  vim cli_notes.md
  146  cat cli_notes.md
  147  cat notes.txt | sort
  148  cat cli_notes.md
  149  cat notes.txt
  150  cat notes.txt | sort 
  151  cat notes.txt | sort > sorted_notes.txt
  152  cat notes.txt | grep Mil
  153  cat notes.txt | grep week01
  154  cd
  155  mkdir gitfoldernew
  156  ls
  157  git init
  158  git status
  159  rm -rf .git
  160  cd gitfoldernew
  161  git init
  162  git status
  163  touch newgitfile
  164  git status
  165  git add new gitfile
  166  git add newgitfile
  167  git status
  168  git commit -m "my first git commit"
  169  git status
  170  mkdir take2wgit
  171  cd take2wgit
  172  git init
  173  touch take1
  174  git add take1
  175  git remote add origin https://github.com/torrepaul315/firstrepositoryinclass.git
  176  git push -u origin master
  177  git push gitfoldernew/
  178  git status
  179  git commit take1
  180  git status
  181  git remote - v
  182  git remote-v
  183  git remote -v
  184  git push - origin master
  185  git push origin master
  186  git push orgin master
  187  git checkout -v
  188  git checkout -b test
  189  git status
  190  git branch
  191  git merge master
  192  git clone
  193  git clone gitfoldernew
  194  git clone take2wgit
  195  git push origin master
  196  rm -rf .git
  197  cd ..
  198  ls -a
  199  rm -rf .git
  200  git status
  201  git clone https://github.com/torrepaul315/git-branch-exercise.git
  202  ls
  203  cd git-branch-exercise

routes/index.js
var express = require('express');
var router = express.Router();
var knex = require('../db/knex');

/* GET home page. */
router.get('/blogpost', function(req, res, next) {
   knex('blogpost').orderBy('blogpost_timestamp')

  // res.render('index', { title: 'Express' });
  .then(blogpost => {
    res.send(blogpost);
  })
});

router.get('/blogpost/:id', (req, res, next) => {
  knex('blogpost').where('id', req.params.id)
  .then(blogpost => {
    res.send(blogpost);
  })
})

// have it set up, but probably won't need it!
router.get('/user', (req,res,next) => {
  knex('user')

  .then(user => {
    res.send(user);
  })
})
//how I pull in author info when I use multiple gets
router.get('/user/:email', (req, res, next) => {
  knex('user').where('email', req.params.email)
  .then(user => {
    res.send(user);
  })
})
// get all comments
router.get('/comment', (req,res,next) => {
  knex('comment')

  .then(comment => {
    res.send(comment);
  })
})
//get comments tied to a blogpost
router.get('/comment/:id', (req,res,next) => {
  knex('comment').where('blogpost_id', req.params.id)
  .then(comment=> {
    res.send(comment);
  })
  .catch(err => {
    res.status(503).send(err.message)
  })
})
//where you grab the info from when you're going to the edit comment page

router.get('/comment/edit/:id', (req,res,next) => {
  knex('comment').where('id', req.params.id)
  .then(comment=> {
    res.send(comment);
  })
  .catch(err => {
    res.status(503).send(err.message)
  })
})
//where I go out a and use a join to grab a comment AND the user email/name!

router.get('/comment/join/:id', (req,res,next) => {
  knex('comment')
  .join('user', 'comment.user_email', '=','user.email')
  .select('*')
  .where('id', req.params.id)
  .then(joinTable => {
    res.send(joinTable)
  }).catch(err => {
    res.status(503).send(err.message)
  })
})


// knex('users')
// .join('contacts', 'users.id', '=', 'contacts.user_id')
// .select('users.id', 'contacts.phone')
// Outputs:
// select `users`.`id`, `contacts`.`phone` from `users` inner join `contacts` on `users`.`id` = `contacts`.`user_id`



//remember! http POST localhost:8000/zebras name="fred" location="San diago" stripes:=8 if the data is a string ="" if an int, :=8
//p 2 - as I go further down the road, leveraging the example of movie crud, should I refactor and prep that object outside of the routes page? A - could, but shouldn't need to


//as per brad- with the routes....you need to first go to the authors list, to see if there's an author in the database....if yes, then use that existing id in the post....if NOT, then insert into the user/authors table that person,, return the id (wierd .returning('id').)


//first attempt at getting info in ergo, submitting a user email string- this will need to be altered based on info from brad


//ao 9:30- the basic functionality of post works!



/*! on your blogpost page, you'll need to take the name and email*/
// ergo, you'll also need to test the post calls feeding the router that info!



//walking through the logic here
router.post('/blogpost', (req,res,next) => {
  knex('user').where('email', req.body.user_email).first()
  .then(user => {
    if (!user) {
      return knex('user')
        .insert({
          name: req.body.name,
          email:req.body.user_email,
        })
    }
  })
  .then(()=> {
  return knex('blogpost').insert({
  title:req.body.title,
  body:req.body.body,
  user_email: req.body.user_email,
  } ,'id')
  })
  .then(id => {
    res.status(200).send(`something happened...${id}`)
  })
  .catch(err => {
    res.status(503).send(err.message)
  })
})
//figure out the post/patch/delete routes

// router.put()



//as per danny, this is another fairly complicated post request-
//1- need to check if user exists (much like the blogpost post req) 2- you also need to check if the blogpost commented on exists!
//not entirely sure if you will need to check if the blogpost exists.....the edit comment page could have a large form much like the edit form on movies and therefore workaround the need for a second .then statement!
router.post('/comment', (req,res,next) => {
  knex('user').where('email', req.body.user_email).first()
  .then(user => {
    if (!user) {
      return knex('user')
        .insert({
          name: req.body.name,
          email:req.body.user_email,
        })
    }
  })
  .then(() => {
  return knex('comment').insert({
    body:req.body.body,
    user_email:req.body.user_email,
    blogpost_id:req.body.blogpost_id,
    /* the email can be submitted because it can be scrubbed, but a second loop to .then check for a blogpost id seems like something that could be caught more upstream!*/
    //  table.integer('blogpost_id').references('id').inTable('blogpost')
    }, 'id')
  })
  .then(id => {
    res.send(`something else happened ${id}`)
  })


})

/* will need
router.put route to edit a blogpost,
and a
router.put to edit a comment
*/

router.put('/blogpost/:id', (req,res,next) => {
  knex('blogpost').where('id', parseInt(req.params.id)).update({
  title: req.body.title,
  body: req.body.body,
  }, 'id').then(id => {
  res.send(`something else happened ${id}`)
  });
})
//about this route...I'm rendering the body and author, but only really letting them edit the body? hmm
router.put('/comment/:id', (req,res,next) => {
  knex('comment').where('id', parseInt(req.params.id)).update({
  body:req.body.body,

}, 'id').then(id => {
    res.send(`something else happened ${id}`)
  })
})
// table.increments('id').primary();
// table.text('body');
// table.timestamp('comment_timestamp').defaultTo(knex.fn.now());
// table.string('user_email').references('email').inTable('user');
// table.integer('blogpost_id').references('id').inTable('blogpost')
// });

router.delete('/blogpost/:id', (req,res,next) => {
  knex('comment').where('blogpost_id', parseInt(req.params.id)).del()

  .then(() => {
  return knex('blogpost').where('id', parseInt(req.params.id)).del()
  })
  .then(() => {
    res.status(204).send()
  })
  .catch(err => {
    res.status(503).send(err.message)
  })
})



router.delete('/comment/:id', (req,res,next) => {
  knex('comment').where('id',
  parseInt(req.params.id)).del()
  .then(() => {
    res.status(204).send()
  });
})


module.exports = router;
